linkerd/main/src/main/scala/io/buoyant/linkerd/TerminationHook.scala

```diff
@@ -0,0 +1,27 @@
+package io.buoyant.linkerd
+
+import com.twitter.util.Duration
+import sun.misc.{Signal, SignalHandler}
+
+trait TerminationHook {
+
+  /**
+   * Trap termination signals and triggers an App.close for a graceful shutdown.
+   * Shutdown hook is not used because it has, at least, the following problems:
+   * <ul>
+   *   <li>LogManager uses a shutdown hook which makes nothing to be logged during shutdown
+   *   <li>TracerCache uses a shutdown hook to flush
+   * </ul>
+   */
+  def register(shutdown: Duration => Unit, shutdownGrace: Duration): Unit = {
```

Comment:
I've got some concerns about this trait but I think it can just go away entirely if the refactor that I objected to above is reverted.